### PR TITLE
Multicoloured spesos: add layer function in StackComponent

### PR DESCRIPTION
**Pull request by whatston3**
## About the PR

Spesos should now appear in non-single denominations.

StackComponents now have an optional layerFunction that will adjust the layer data before it is used in the StackVisualizer.

This is only used in spesos stacks at the moment, with one concrete instance that uses threshold values (used with bill denominations for the purpose of spesos).

## Why / Balance

Looks pretty, feels better when you make more money.

My biggest concerns here are whether this is fragile with respect to upstream changes, and if the names of things make sense.

## How to test

Build and run local server.
1. Spawn in spesos in $1, $10, $100, $500, $1000, and $1000000 denominations.
2. Try merging and splitting 1 off from all bills larger than $1.

## Media

spesos

- [X] ***I have added screenshots/videos to this PR showcasing its changes ingame***, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- fix: Stacks of spesos now display the largest valued bill on top.

### CODE REVIEW

**Comment by dvir001** [comment]
Can you also cap the stack to max $1000000 / 100000 ?
Might as well since its a bit weird you can have infinite amounts on a single stack

**Comment by whatston3** [comment]
> Can you also cap the stack to max $1000000 / 100000 ? Might as well since its a bit weird you can have infinite amounts on a single stack

I'll look into how this works with cargo sales and ATM withdrawals, not the largest fan of the idea, makes handling large transfers (e.g. spessman's lotto) awkward.

**Comment by Leander-0** [comment]
I like the idea of the money being stacked, we can add money bags that act as the mining bags to handle large amounts of money. kinda silly that one can carry 1 million spessos on their pocket without issue. also adds gameplay to hide the money from pirates.

**Comment by whatston3** [comment]
> I'll look into how this works with cargo sales and ATM withdrawals, not the largest fan of the idea, makes handling large transfers (e.g. spessman's lotto) awkward.

Gonna say this is out of the scope of the PR.  Capping the max stack size causes issues with ATM withdrawal ("admin, help, I asked for 5 million spesos, ATM gave me 1 and ate the other 4").  What I'd argue _is_ in scope is a few new bill denominations (10K, 25K, 50K, and 100K all scroll - space on bill face for amount is tiny - but do not change colour).  Pink bill needs rework, my spriteing is awful.

image

**Comment by dvir001** [comment]
> > I'll look into how this works with cargo sales and ATM withdrawals, not the largest fan of the idea, makes handling large transfers (e.g. spessman's lotto) awkward.
> 
> Gonna say this is out of the scope of the PR. Capping the max stack size causes issues with ATM withdrawal ("admin, help, I asked for 5 million spesos, ATM gave me 1 and ate the other 4"). What I'd argue _is_ in scope is a few new bill denominations (10K, 25K, 50K, and 100K all scroll - space on bill face for amount is tiny - but do not change colour). Pink bill needs rework, my spriteing is awful.
> 
> image

Sound good

**Comment by whatston3** [comment]
Adjusted style of large bills:
* all scroll left
* all use 4-pixel wide zeros
* colours avoid repetition as best as possible
* pink bill changed to a bright red [still probably too bright]
* rainbow bill flashing follows order of currency, uses buff colour at bottom of digits
* rainbow bill value changed to 250k (rationale here: if we will have a maximum size at some point, you shouldn't be able to hide your flashing bills easily by splitting it easily

Preview of bills as they stand (bottom row are animated - 10k, 25k, 50k, 100k, 250k+)
image

File: Content.Shared/Stacks/StackComponent.cs [hunk 1]
```diff
@@ -78,6 +78,14 @@ public sealed partial class StackComponent : Component
         [DataField("layerStates")]
         [ViewVariables(VVAccess.ReadWrite)]
         public List<string> LayerStates = new();
+
+        // Frontier: transforming Amount, MaxCount in speso stacks 
+        /// <summary>
+        /// An optional function to adjust the layers used for a stack's appearance.
+        /// </summary>
+        [DataField("layerFunction")]
+        public StackLayerFunction? LayerFunction;
```
Comment (Cheackraze) [inline Content.Shared/Stacks/StackComponent.cs:87]: hm this seems really smelly to me. It seems like you're storing a function into a component, which isnt serializable and will break any kind of saving/loading or serialization. The benefits of keeping components to only data, and a fully ECS system.

File: Content.Client/Stack/StackSystem.cs [hunk 2]
```diff
@@ -56,20 +56,25 @@ private void OnAppearanceChange(EntityUid uid, StackComponent comp, ref Appearan
...
+            if (!_appearanceSystem.TryGetData<int>(uid, StackVisuals.MaxCount, out data.MaxCount, args.Component))
+                data.MaxCount = comp.LayerStates.Count;
+
+            if (!_appearanceSystem.TryGetData<bool>(uid, StackVisuals.Hide, out data.Hidden, args.Component))
+                data.Hidden = false;
 
-            if (!_appearanceSystem.TryGetData<bool>(uid, StackVisuals.Hide, out var hidden, args.Component))
-                hidden = false;
+            if (comp.LayerFunction is not null) // Frontier: use stack layer function to modify appearance if provided.
+                comp.LayerFunction.Apply(ref data); // Frontier
```
Comment (Cheackraze) [inline Content.Client/Stack/StackSystem.cs:72]: instead, you should just add a bool flag to the stack component to check for that instead, and then call the shared function here after bringing it in as a dependency.

**Review by Cheackraze — CHANGES_REQUESTED** [review]

**Comment by whatston3** [comment]
Will review this with fresher eyes before requesting another review, but @Cheackraze, your concerns should be addressed.

**Comment by whatston3** [comment]
Changes should be addressed.  Still functional, and if you have a better name for `ThresholdStackLayerFunctionComponent`, I'm all ears.

**Comment by Cheackraze** [comment]
> Changes should be addressed. Still functional, and if you have a better name for `ThresholdStackLayerFunctionComponent`, I'm all ears.

at is a bit wordy for what it is. Its just a component, its use for a function is irrelevant. maybe just, 'StackLayerThresholdComponent'

**Comment by whatston3** [comment]
> at is a bit wordy for what it is. Its just a component, its use for a function is irrelevant. maybe just, 'StackLayerThresholdComponent'

Sounds good to me.  Renamed file, reworded a comment, adjusted component name in speso YAML definition.  Still works alright, no issues.

**Review by Cheackraze — APPROVED** [review]
